src/lib.rs

```diff
@@ -44,13 +39,102 @@ pub fn hexf32(input: TokenStream) -> TokenStream {
 /// ```
 #[proc_macro]
 pub fn hexf64(input: TokenStream) -> TokenStream {
-    let lit = syn::parse_macro_input!(input as syn::LitStr);
-    match hexf_parse::parse_hexf64(&lit.value(), true) {
-        Ok(v) => format!("{:?}f64", v) // should keep the sign even for -0.0
-            .parse()
-            .expect("formatted a f64 literal"),
-        Err(e) => syn::Error::new(lit.span(), format!("hexf64! failed: {}", e))
-            .to_compile_error()
-            .into(),
+    parse_stream(input, |s, span| match hexf_parse::parse_hexf64(s, true) {
+        Ok(v) => TokenTree::Literal(Literal::f64_suffixed(v)).into(),
+        Err(e) => compile_error_stream(&format!("hexf64! failed: {}", e), span),
+    })
+}
+
+const EXPECTED_STRING_LITERAL: &str = "expected string literal";
+
+fn parse_stream<F>(input: TokenStream, process_literal: F) -> TokenStream
+where
+    F: FnOnce(&str, Span) -> TokenStream,
+{
+    let literal = match get_literal(input) {
+        Ok(lit) => lit,
+        Err(tokens) => return tokens,
+    };
+
+    match extract_string_value(&format!("{}", literal)) {
+        Some(s) => process_literal(s, literal.span()),
+        None => compile_error_stream(EXPECTED_STRING_LITERAL, literal.span()),
+    }
+}
+
+fn get_literal(input: TokenStream) -> Result<Literal, TokenStream> {
+    let mut trees = input.into_iter();
+    let tree = trees.next().ok_or_else(|| {
+        compile_error_stream(
+            "unexpected end of input, expected string literal",
+            Span::call_site(),
+        )
+    })?;
+
+    if let Some(next_tree) = trees.next() {
+        return Err(compile_error_stream("unexpected token", next_tree.span()));
+    }
+
+    match tree {
+        TokenTree::Literal(literal) => Ok(literal),
+        _ => Err(compile_error_stream(EXPECTED_STRING_LITERAL, tree.span())),
     }
 }
+
+fn extract_string_value(literal: &str) -> Option<&str> {
+    if literal.starts_with('"') {
+        if literal.len() >= 2 && literal.ends_with('"') {
+            // For simplicity, we do not handle any escape sequences. Given the
+            // usage of these macros, they would be of questionable utility.
```

Comment:
I agree we shouldn't handle them here, but I think we should clearly indicate that escape sequences are explicitly not supported. I'm okay to reject the string if its code contains `\`. The same comment applies to the other branch.

Reply:
Added a specific error message when escape sequences are detected.